codeworld-server/src/Util.hs

```diff
@@ -296,3 +299,53 @@ removeDirectoryIfExists dirName =
     handleExists e
         | isDoesNotExistError e = return ()
         | otherwise = throwIO e
+
+loadDumpedTree :: BuildMode -> UserId -> IO DirTree
+loadDumpedTree bm uid = do
+    let file = userProjectDir bm uid </> "tree.info"
+    exists <- doesFileExist file
+    realTree <- getDirectoryTree bm uid
+    case exists of
+        False -> return realTree
+        True -> do
+            content  <- LB.readFile file
+            return $ case decode content of
+                Nothing -> realTree
+                Just dumpedTree -> case compareTrees realTree (Dir "root" dumpedTree) of
+                    True  -> Dir "root" dumpedTree
+                    False -> realTree
+
+getDirectoryTree :: BuildMode -> UserId -> IO DirTree
+getDirectoryTree bm uid = do
+    ftr <- getDirectory $ userProjectDir bm uid
```

Comment:
I'm not clear on what `ftr` means.  More descriptive name, please?

Reply:
I'll rename it to `fileTree`